wallet-new/src/Cardano/Wallet/Kernel/Migration.hs

```diff
@@ -0,0 +1,229 @@
+    case resolved of
+       Nothing -> -- We assume no migration is needed and we move along
+           logMsg Info $ "No legacy DB at " <> pack unresolvedDbPath <> " , migration is not needed."
+       Just legacyDbPath -> do
+           bracketLegacyDB legacyDbPath $ \st -> do
+                let  (unavailable, available) = partitionEithers (metadataFromWalletStorage st)
+                     unavailableLen = length unavailable
+                     availableLen   = length available
+
+                when (unavailableLen > 0) $ do
+                    logMsg Error $ show unavailableLen
+                        <> " out of "
+                        <> show unavailableLen
+                        <> " rootAddress(es) failed to decode"
+
+                when (availableLen > 0) $ do
+                    logMsg Info $ "Found "
+                        <> show availableLen
+                        <> " rootAddress(es) to migrate."
+
+                mapM_ (restore logMsg pw keystore) available
+
+           -- Now that we have closed the DB, we can move the directory
+           backupPath <- moveLegacyDB legacyDbPath
+
+           -- asynchronous restoration still runs at this point.
+           logMsg Info $ "acid state migration succeeded. Old db backup can be found at " <> pack backupPath
+
+
+restore :: (Severity -> Text -> IO ())
+        -> Kernel.PassiveWallet
+        -> Keystore
+        -> MigrationMetadata
+        -> IO ()
+restore logMsg pw keystore metadata = do
+    let wId = WalletIdHdRnd (metadata ^. mmHdRoot . HD.hdRootId)
+        -- Our best guess whether or not this wallet has the spending password.
+        -- currently we always yield True, but we can tweak these values if
+        -- our assumptions reveal not to be correct (or if we can fine tune
+        -- our heuristic).
+        bestGuess = case metadata ^. mmHasSpendingPasswordHeuristic of
+                         LowConfidence  -> True
+                         HighConfidence -> True
+    mEsk <- Keystore.lookup wId keystore
+    case mEsk of
+        Just esk -> do
+            res <- restoreWallet pw
+                                 bestGuess
+                                 (metadata ^. mmDefaultAddress)
+                                 (metadata ^. mmHdRoot . HD.hdRootName)
+                                 (metadata ^. mmHdRoot . HD.hdRootAssurance)
+                                 esk
+                                 (prefilter esk pw wId)
+            case res of
+                 Right (restoredRoot, balance) -> do
+                     let errMsg = "Migrating " % F.build
+                                % " with balance " % F.build
+                     logMsg Info (F.sformat errMsg restoredRoot balance)
+                 Left err -> do
+                     let errMsg = "Couldn't migrate " % F.build
+                                % " due to : " % F.build
+                     logMsg Error (F.sformat errMsg wId err)
```

Comment:
If we go with the `----must-restore-all-wallets` argument, then we have a bit of a diffculty here; if we now wanted to restart the migration, the user must manually delete the (new) DB and try again.

Reply:
Should probably document this in the migration instructions.